File: go/dsa/algorithms/sliding-window/min-window.go
func minWindow(s string, t string) string {
    if len(s) < len(t) {
        return ""
    }

    targetFreq := make(map[byte]int)
    for i := 0; i < len(t); i++ {
        targetFreq[t[i]]++
    }

    requiredMatches := len(targetFreq)
    windowFreq := make(map[byte]int)
    left, matchCount := 0, 0
    minLength, bestStart := 1<<31-1, 0

    for right := 0; right < len(s); right++ {
        // 1️⃣ Expand
        c := s[right]
        if count, ok := targetFreq[c]; ok {
            windowFreq[c]++
            if windowFreq[c] == count {
                matchCount++
            }
        }

        // 2️⃣ Condition: while valid
        for matchCount == requiredMatches {
            // 4️⃣ Record / Optimize
            if right - left + 1 < minLength {
                minLength = right - left + 1
                bestStart = left
            }

            // 3️⃣ Contract
            leftChar := s[left]
            if count, ok := targetFreq[leftChar]; ok {
                if windowFreq[leftChar] == count {
                    matchCount--
                }
                windowFreq[leftChar]--
            }
            left++
        }
    }

    if minLength == 1<<31-1 {
        return ""
    }
    return s[bestStart : bestStart+minLength]
}

File: cpp/dsa/algorithms/sliding-window/min-window.cpp
string minWindow(string s, string t) {
    if (s.length() < t.length()) return "";

    unordered_map<char, int> targetFreq;
    for (char c : t) {
        targetFreq[c]++;
    }

    int requiredMatches = targetFreq.size();
    unordered_map<char, int> windowFreq;
    int left = 0, matchCount = 0;
    int minLength = INT_MAX, bestStart = 0;

    for (int right = 0; right < s.length(); right++) {
        // 1️⃣ Expand
        char c = s[right];
        if (targetFreq.count(c)) {
            windowFreq[c]++;
            if (windowFreq[c] == targetFreq[c]) {
                matchCount++;
            }
        }

        // 2️⃣ Condition: while valid
        while (matchCount == requiredMatches) {
            // 4️⃣ Record / Optimize
            if (right - left + 1 < minLength) {
                minLength = right - left + 1;
                bestStart = left;
            }

            // 3️⃣ Contract
            char leftChar = s[left];
            if (targetFreq.count(leftChar)) {
                if (windowFreq[leftChar] == targetFreq[leftChar]) {
                    matchCount--;
                }
                windowFreq[leftChar]--;
            }
            left++;
        }
    }

    return minLength == INT_MAX ? "" : s.substr(bestStart, minLength);
}